### PR TITLE
EAMxx: add binary ops diags

### PR DESCRIPTION
**Pull request by mahf708**
add some binary arithmetics for EAMxx diags

[bfb]

### CODE REVIEW

**Review by bartgol — APPROVED** [review]
If this is needed soon-ish, then this is ok. I think long term we may want (or maybe not, up for debate) a more broad/robust approach. In particular, I think people may want to compute semi-arbitrary expressions, in which case we may replace this with a more generic "expression-parser" kind of diagnostic (perhaps incorporating some existing TPL).

I also recommend that this be merged after the output field aliasing PR, so we can rename fields in the nc file and avoid +-*/ symbols in var names.

File: components/eamxx/src/share/io/eamxx_io_utils.cpp [hunk 1]
```diff
@@ -141,6 +141,7 @@ create_diagnostic (const std::string& diag_field_name,
   std::regex horiz_avg ("([A-Za-z0-9_.]+)_horiz_avg$");
   std::regex vert_contract ("([A-Za-z0-9_.]+)_vert_(avg|sum)(_((dp|dz)_weighted))?$");
   std::regex zonal_avg (R"(([A-Za-z0-9_.]+)_zonal_avg_(\d+)_bins$)");
+  std::regex binary_ops (R"(([A-Za-z0-9_.]+)_(\+|\-|\*|\÷)_([A-Za-z0-9_.]+)$)");
```
Comment (mahf708) [inline components/eamxx/src/share/io/eamxx_io_utils.cpp]: @bartgol is this syntax ok with you? A_*_B (note the underscore)

Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: Also, we need to add *,+,-,/ to this `([A-Za-z0-9_.]+)`... ugh

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: Oh, I did not notice that. It seems a bit ugly. Can we support maybe just `A*B`? Bonus points if we can allow spaces, for readability, like `A * B`.

Another thought: this would prob not work in case of `A*B+C`, right? I think the current piping of diags may group this as "A times (B+C)". Maybe we should quickly verify if this is right, and warn users against compounding these expressions in the docs?

Edit: regardless of how our piping works, i think one between A*B+C and A+B*C would be evaluated incorrectly, which means we need to warn users about mixing */ with +-.

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: Speaking of, maybe we should add an entry to the docs?

Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: > Edit: regardless of how our piping works, i think one between AB+C and A+BC would be evaluated incorrectly, which means we need to warn users about mixing */ with +-.

Yes the piping will work incorrectly as it will eval them from the back to the front.

Docs are coming (including caveats)

> Oh, I did not notice that. It seems a bit ugly

yeah, that's why I am asking. I made this decision because: `field_at_500hPa*field_horiz_avg` looks less coherent that field_at_500hPa_*_field_horiz_avg. Same for `T_mid*p_mid` vs `T_mid_*_p_mid`, but I am happy with either one.

I don't like allowing spaces because it will become inconsistent at some point (like do we allow `field_horiz_avg` AND `field horiz_avg`...?

Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: (I think spaces for `:=` is a good idea as it is a special separator of sorts)

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: I thought our IO layer would already strip names from any space, if any. Actually, I think the yaml parser does not even see leading/trailing spaces in lists (so `a: [1,2,3]` is the same as `a: [1,  2  , 3]`). I _think_.

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: We can ask other ppl, but imho `T_mid*p_mid` is easier to read than `T_mid_*_p_mid`.

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: Ah, but the `-` sign may conflict with the sign for a negative number. I suppose if we consider binary_ops as the last resort, we already ruled out the "blah_where_X_le_-20", so maybe that's ok.

I feel like we need to sit down and see which syntaxes need to be rethought...

Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp]: For now, we could err on the cautious (and ugly) side... but yeah, I am planning to put a HARD STOP on these diags after the currently open PRs. No more hacks for accomodation. Next big round will likely have to be a big rethought from the gounds up...

File: components/eamxx/docs/user/diags/binary_ops.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+1. **Same layout**
+2. **Same data type**
+3. **Same grid**
+4. **Compatible units** for addition and subtraction
+
+## Unit handling
+
+The resulting diagnostic field will have units determined by the operation.
+
+## Configuration
```
Comment (bartgol) [inline components/eamxx/docs/user/diags/binary_ops.md:31]: Now that it is available, should we remind the user about aliasing, and "encourage" to use it for binary-op diags?

Comment (mahf708) [reply components/eamxx/docs/user/diags/binary_ops.md:31]: no, I would keep the docs for each feature separate. I will write an overarching docs page for tips to use them together, but I think each page should be isolated to what it is offering as a standalone feature

File: components/eamxx/src/share/io/eamxx_io_utils.cpp [hunk 3]
```diff
@@ -129,21 +129,25 @@ create_diagnostic (const std::string& diag_field_name,
   //       of each group in the matches output var!
   // Note: use raw string syntax R"(<string>)" to avoid having to escape the \ character
   // Note: the number for field_at_p/h can match positive integer/floating-point numbers
-  std::regex field_at_l (R"(([A-Za-z0-9_.+-]+)_at_(lev_(\d+)|model_(top|bot))$)");
-  std::regex field_at_p (R"(([A-Za-z0-9_.+-]+)_at_(\d+(\.\d+)?)(hPa|mb|Pa)$)");
-  std::regex field_at_h (R"(([A-Za-z0-9_.+-]+)_at_(\d+(\.\d+)?)(m)_above_(sealevel|surface)$)");
+  // Start with a generic for a field name allowing for all letters, all numbers, dash, dot, plus, minus, product, and division
```
Comment (mahf708) [inline components/eamxx/src/share/io/eamxx_io_utils.cpp:132]: @bartgol pls re-review this file

File: components/eamxx/src/share/io/eamxx_io_utils.cpp [hunk 4]
```diff
@@ -129,21 +129,25 @@ create_diagnostic (const std::string& diag_field_name,
...
+  // Escaping all the special ones just in case
+  std::string generic_field = "([A-Za-z0-9_.+\\-\\*\\÷]+)"; 
+  std::regex field_at_l (R"()" + generic_field + R"(_at_(lev_(\d+)|model_(top|bot))$)");
```
Comment (bartgol) [inline components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: What's the leading "()" for?

**Review by bartgol — APPROVED** [review]

File: components/eamxx/src/share/io/eamxx_io_utils.cpp (hunk 4 above)
Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: To ensure the std::string gets converted to the proper stuff? Would it work without it?  I didn't test ... that's why I wanted you to review this before merging 😅

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: std::regex constructor takes a string, and generic_field is already a std::string (otherwise the + operator would not work). The reason why we need the R in other  strings is to prevent stuff like `\n` to be escaped. And speaking of, you can just do

```c++
std::string generic_field = "([A-Za-z0-9_.+-*÷]+)"; 
```
no need to escape characaters inside `[]`. And since you don't have fancy chars, no need for R. Btw, why `÷` instead of `/`? It's not a standard char, so maybe not what ppl would use...

File: components/eamxx/docs/user/diags/binary_ops.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+## Example
+
+```yaml
+field_names:
+  - T_mid_*_p_mid
+```
+
+## Caveats
+
+- Strictly speaking, multiple operations will take place in order.
```
Comment (AaronDonahue) [inline components/eamxx/docs/user/diags/binary_ops.md:56]: Should we note that we don't support parentheses? So no `_(_qc_+_qv_)_*_p_mid`

Comment (AaronDonahue) [reply components/eamxx/docs/user/diags/binary_ops.md:56]: Second question, do we have a test for this case?  I may have missed it, but I think the test just does binary ops for 2 vars.  It seems like 3 vars is a good edge case, and sense you mention it as an example it would be good to test it.

File: components/eamxx/src/share/io/eamxx_io_utils.cpp (hunk 4 above)
Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: - Not escaping `*` and/or `÷` (or maybe it was having stuff after un-escaped `-`) caused issues in other tests (io_remap, io_diags, etc.). I can bring it back to see the errors if you want...
- `/` is not supported in netcdf writing; `÷` is

File: components/eamxx/docs/user/diags/binary_ops.md (hunk 5 above)
Comment (mahf708) [reply components/eamxx/docs/user/diags/binary_ops.md:56]: Yes, both are good suggestions. I can include them or in a follow-up PR (that will combine docs for all features).

I will also need to clarify very clearly the order of ops by giving a few examples...

File: components/eamxx/src/share/io/eamxx_io_utils.cpp (hunk 4 above)
Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: Uhm, `÷` is not ASCII though, which may be undesirable...

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: I wonder if we should replace symbols with "words" in var names. E.g., 
- yaml request `f_1+f_2` saved as `f_1_plus_f_2`
- yaml request `f_1-f_2` saved as `f_1_minus_f_2`
- yaml request `f_1*f_2` saved as `f_1_times_f_2`
- yaml request `f_1/f_2` saved as `f_1_over_f_2`

Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: > * yaml request `f_1+f_2` saved as `f_1_plus_f_2`
> * yaml request `f_1-f_2` saved as `f_1_minus_f_2`
> * yaml request `f_1*f_2` saved as `f_1_times_f_2`
> * yaml request `f_1/f_2` saved as `f_1_over_f_2`

I am not a fan of changing what the request is. 

I am okay with `f_1_plus_f_2`, etc. --- should I just switch that? I don't wanna overcomplicate stuff, so we should just pick one or another. The plus, minus, times, over terminology is ok. We can also do, addop, subop, prodop, divop etc.

Comment (bartgol) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: So you want the name in the yaml to match the name in the nc file? That's reasonable. I don't have a strong pref between addop and plus, so you can pick there. But I think sticking with words may be best long term (I am already fearing some postproc script choking on a + or -...)

Comment (mahf708) [reply components/eamxx/src/share/io/eamxx_io_utils.cpp:135]: yeah, unlesss someone asks for the alias feature (also remember, in the aliasing feature, we document the name in the attrs). Already, will ping you on slack for one last call on plus vs addop

File: components/eamxx/docs/user/diags/binary_ops.md (hunk 5 above)
Comment (mahf708) [reply components/eamxx/docs/user/diags/binary_ops.md:56]: @AaronDonahue, I will add the composability test later. The composability comes from the parsing (in eamxx IO utils), not the diag impl itself. This PR is mostly about the diag impl. I added some notes, but I will strengthen the testing later.

File: components/eamxx/docs/user/diags/binary_ops.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+- As the name suggests, these diagnostics were written for a single
+  operation connecting two fields; doing multiple binary ops in
+  succession carries risks; read on.
+- Strictly speaking, multiple operations will take place in order.
+  For example, `qc_plus_qv_times_p_mid` will first compute `qc + qv`,
+  then multiply the result by `p_mid`. (Beware, we do NOT support
+  mathematical operation precedence; it is simply about the parser
+  which processes the expression left-to-right.)
+- In fact, `p_mid_times_qc_plus_qv` will fail because of units!
+- We only support existing fields in the Field Manager, e.g.,
```
Comment (bartgol) [inline components/eamxx/docs/user/diags/binary_ops.md:62]: Maybe 
```yaml
We only support existing fields in the Field Manager (and other diagnostics). e.g. ...
```

**Review by bartgol — APPROVED** [review]
Looks good to go. I had a docs comment, but I think you're going to work on a diags (docs) overhaul, so it can wait.

A thought: it'd be nice if we allowed one of the two operand to be a known constant. E.g., one could request `f1_times_rho_w` or something like that. But we can think about this later.
